ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/QueryProjectNode.java

```diff
@@ -167,6 +170,17 @@ private LogicalSchema buildOutputSchema(final MetaStore metaStore) {
 
       outputSchema = selectOutputSchema(metaStore, projects, isWindowed);
     }
+
+    if (isScalablePush) {
+      // Transient queries return key columns in the value, so the projection includes them, and
+      // the schema needs to include them too:
```

Comment:
I would say that the primary reason is that we use the same physical layer to execute the operation as with pull queries (`ProjectOperator`), and this assumes it's using a `QueryProjectNode`.  I agree that this is a bit of redundant logic since this class exists largely because pull queries are not handled well by `FinalProjectNode` -- see comments here: https://github.com/confluentinc/ksql/pull/6684

For SPQs, it might be worth detangling these things.  We also make use of some additional things like the concept of an intermediate schema.  I think the intention was that we could push down projections and do other things that are normally done in a relational db, where each layer might have a slightly different schema required to do the computation.  This is a different concept from how persistent and transient queries work today (and we currently don't do the optimization since we're passing around a table row with the whole data anyway 🤦 ).  We'll have to eventually ask Vicky what the vision there is.  Anyway, I could see us eventually merging back `QueryProjectNode` and `FinalProjectNode`, or diverging and making `QueryProjectNode` more specialized for relational type operations.

Reply:
ack